include/range/v3/utility/semiregular.hpp

```diff
@@ -18,68 +18,126 @@
 #include <meta/meta.hpp>
 #include <range/v3/range_fwd.hpp>
 #include <range/v3/utility/get.hpp>
-#include <range/v3/detail/optional.hpp>
+#include <range/v3/utility/optional.hpp>
 
 namespace ranges
 {
     inline namespace v3
     {
         /// \addtogroup group-utility
         /// @{
-        template<typename T>
-        struct semiregular
+
+        /// \cond
+        namespace detail
         {
-        private:
-            optional<T> t_;
-        public:
-            semiregular() = default;
-            semiregular(T f)
-              : t_(std::move(f))
-            {}
-            template<typename ...Args>
-            semiregular(in_place_t, Args &&...args)
-              : t_(in_place, static_cast<Args&&>(args)...)
-            {}
-            T & get()
+            template<typename T>
+            struct semiregular_base
```

Comment:
If you move the contents of `semiregular_base` into `semiregular`, you can eliminate `semiregular_base` from the design altogether and use `optional<T>` directly.